<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Hierarchical TAD &#8212; TADLib 0.3.1 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Hierarchical TAD Identification" href="hitad_api.html" />
    <link rel="prev" title="Hi-C Data Loading" href="calfea_api.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">TADLib</a></h1>



<p class="blurb">A Library to Explore Chromatin Interaction Patterns for Topologically Associating Domains</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=XiaoTaoWang&repo=TADLib&type=star&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Install TADLib</a></li>
<li class="toctree-l1"><a class="reference internal" href="calfea.html">Aggregation Preference</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Hierarchical TAD</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#snapshot">Snapshot</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tutorial">Tutorial</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#command-line-tool">Command-line tool</a></li>
<li class="toctree-l3"><a class="reference internal" href="#domain-alignment">Domain Alignment</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#api-documentation">API Documentation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="hitad_api.html">Hierarchical TAD Identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="hitad_api.html#domain-loading-and-aligning">Domain Loading and Aligning</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Release Notes</a></li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="hierarchical-tad">
<h1>Hierarchical TAD<a class="headerlink" href="#hierarchical-tad" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>HiTAD is a method to detect hierarchical TADs, including TADs, sub-TADs and smaller domains.
Except local insulations, HiTAD further constrains TADs as the optimal domains to globally
separate intra-chromosomal interactions. Under objective functions derived from chromatin
interactions, HiTAD adopts an iterative optimization procedure to detect hierarchical TADs.
HiTAD performs well in domain sensitivity, replicate reproducibility and inter cell-type
conservation. Application to human and mouse cell types (including both traditional Hi-C
and in situ Hi-C data sets) reveals that there exist common change types for hierarchical
TADs, which are involved in shaping higher-order compartment, replication timing and
transcriptional regulation.</p>
</div>
<div class="section" id="snapshot">
<h2>Snapshot<a class="headerlink" href="#snapshot" title="Permalink to this headline">¶</a></h2>
<p>This tutorial will cover the two most commonly used tools/submodules in HiTAD:</p>
<ul>
<li><p class="first">hitad</p>
<p>A command-line tool streamlining our 5-step identification algorithm:</p>
<ol class="arabic simple">
<li>Calculate adaptive directionality index (DI) for each bin.</li>
<li>Detect bottom boundaries by 5-state Gaussian mixture Hidden Markov Model using adaptive
DIs as input.</li>
<li>Identify TAD from bottom domains by optimally separating intra-chromosomal interactions
under chromosome-level objective functions.</li>
<li>Recursively identify inner domains under domain-level objective functions to optimally
separate intra-domain interactions.</li>
<li>(Optional but recommended) Perform domain alignment between two replicates and only
the reproducible domains are maintained to generate the final hierarchical TAD list.</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If the two replicates are comparable in sequencing depth, then step 5 can improve
the accuracy and reliability while guaranteeing the sensitivity of called domains; otherwise,
you&#8217;d better merge the replicates into one dataset and step 5 will be skipped in this case.</p>
</div>
</li>
<li><p class="first">aligner</p>
<p>A submodule containing classes and functions for our proposed domain alignment strategy. In
our work, this strategy is used in reproducible domain detecting and change type defining
at both boundary and domain level.</p>
</li>
</ul>
</div>
<div class="section" id="tutorial">
<h2>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h2>
<div class="section" id="command-line-tool">
<h3>Command-line tool<a class="headerlink" href="#command-line-tool" title="Permalink to this headline">¶</a></h3>
<div class="section" id="usage">
<h4>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h4>
<p><code class="docutils literal"><span class="pre">hitad</span> <span class="pre">[options]</span></code></p>
</div>
<div class="section" id="data-preparation">
<h4>Data Preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">¶</a></h4>
<p>You need to perform a standard Hi-C data processing pipeline (mapping, filtering, binning, correcting)
and get the bin-level contacts for HiTAD.</p>
<p>Only intra-chromosomal contacts are allowed and they should be stored chromosome by chromosome in
text files. These files should be named following the pattern <code class="docutils literal"><span class="pre">chr*.txt</span></code> and be placed under the
same folder, where &#8220;*&#8221; indicates the chromosome labels (1,2,...,X,Y for human). Each file should
contain 3 columns (separated by whitespace):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">bin1</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">bin2</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">IF</span><span class="o">&gt;</span>
</pre></div>
</div>
<ul class="simple">
<li>bin1, bin2: bin index on the corresponding chromosome</li>
<li>IF : normalized/corrected interaction frequency</li>
</ul>
</div>
<div class="section" id="learn-hitad-using-the-txt-data">
<h4>Learn HiTAD Using the TXT Data<a class="headerlink" href="#learn-hitad-using-the-txt-data" title="Permalink to this headline">¶</a></h4>
<p>First, change to the example directory and see what the inputs should look like (suppose you are still
in the distribution root directory):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd lib/hitad/data
$ ls -lh
</pre></div>
</div>
<p>There are two sub-directories named &#8220;NPZ&#8221; and &#8220;TXT&#8221;, corresponding to two supported data formats
respectively:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">total</span> <span class="mi">16</span><span class="n">K</span>
<span class="n">drwxr</span><span class="o">-</span><span class="n">x</span><span class="o">---</span> <span class="mi">2</span> <span class="n">xtwang</span> <span class="n">CPeng</span> <span class="mf">4.0</span><span class="n">K</span> <span class="n">May</span>  <span class="mi">3</span> <span class="mi">13</span><span class="p">:</span><span class="mi">36</span> <span class="n">NPZ</span>
<span class="n">drwxr</span><span class="o">-</span><span class="n">x</span><span class="o">---</span> <span class="mi">4</span> <span class="n">xtwang</span> <span class="n">CPeng</span> <span class="mf">4.0</span><span class="n">K</span> <span class="n">May</span>  <span class="mi">3</span> <span class="mi">13</span><span class="p">:</span><span class="mi">36</span> <span class="n">TXT</span>
</pre></div>
</div>
<p>&#8220;NPZ&#8221; is a binary format we define to accelerate contact data loading. You&#8217;ll know it in more details
soon. For now, just pay attention to the &#8220;TXT&#8221;:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd TXT
$ ls -lh *
</pre></div>
</div>
<p>There are two sub-directories and two files:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">-----</span> <span class="mi">1</span> <span class="n">xtwang</span> <span class="n">CPeng</span>   <span class="mi">37</span> <span class="n">May</span>  <span class="mi">3</span> <span class="mi">13</span><span class="p">:</span><span class="mi">36</span> <span class="n">datasets</span><span class="o">-</span><span class="n">onerep</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">-----</span> <span class="mi">1</span> <span class="n">xtwang</span> <span class="n">CPeng</span>   <span class="mi">65</span> <span class="n">May</span>  <span class="mi">3</span> <span class="mi">13</span><span class="p">:</span><span class="mi">36</span> <span class="n">datasets</span><span class="o">-</span><span class="n">tworeps</span>

<span class="n">IMR90</span><span class="o">-</span><span class="n">HindIII</span><span class="o">-</span><span class="n">rep1</span><span class="p">:</span>
<span class="n">total</span> <span class="mf">2.7</span><span class="n">M</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">-----</span> <span class="mi">1</span> <span class="n">xtwang</span> <span class="n">CPeng</span> <span class="mf">2.7</span><span class="n">M</span> <span class="n">May</span>  <span class="mi">3</span> <span class="mi">13</span><span class="p">:</span><span class="mi">36</span> <span class="n">chr22</span><span class="o">.</span><span class="n">txt</span>

<span class="n">IMR90</span><span class="o">-</span><span class="n">HindIII</span><span class="o">-</span><span class="n">rep2</span><span class="p">:</span>
<span class="n">total</span> <span class="mf">2.8</span><span class="n">M</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">-----</span> <span class="mi">1</span> <span class="n">xtwang</span> <span class="n">CPeng</span> <span class="mf">2.8</span><span class="n">M</span> <span class="n">May</span>  <span class="mi">3</span> <span class="mi">13</span><span class="p">:</span><span class="mi">36</span> <span class="n">chr22</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>The two sub-directories correspond to contact data of two replicates, respectively. And the two
files are metadata files specifying the resolution (in base-pair unit), replicate label and data path
(in TXT case, it should point to the folders) information in a structured way:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cat datasets-tworeps

res:40000
  rep1:../lib/hitad/data/TXT/IMR90-HindIII-rep1
  rep2:../lib/hitad/data/TXT/IMR90-HindIII-rep2

$ cat datasets-onerep

res:40000
  rep1:../lib/hitad/data/TXT/IMR90-HindIII-rep1
</pre></div>
</div>
<p>If you want to use two replicates to improve reliability of identified domains (as we did in our work),
you should provide hitad with a metadata file like &#8220;datasets-tworeps&#8221;; if you have only one replicate
or the sequencing depths between two replicates are quite different, your metadata file should look
like &#8220;datasets-onerep&#8221;.</p>
<p>With contact data and metadata file ready, running hitad is quite straightforward.</p>
<p>Go back to the distribution root and create a new directory there for working:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd ../../../..
$ mkdir working
</pre></div>
</div>
<p>Change to &#8220;working&#8221; directory:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd working
</pre></div>
</div>
<p>Then type in the command below:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ hitad -O IMR90-HindIII-40K-tworeps.txt -d ../lib/hitad/data/TXT/datasets-tworeps --logFile hitad-40K.log
</pre></div>
</div>
<ul>
<li><p class="first"><code class="docutils literal"><span class="pre">-O/--output</span></code> OUTPUT</p>
<p>Output file name.</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">-d/--datasets</span></code> DATASETS</p>
<p>Metadata file path describing your Hi-C data</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--logFile</span></code> LOGFILE</p>
<p>Log file name. (Default: hitad.log)</p>
</li>
</ul>
<p>After this command(it takes about several seconds), two files &#8220;IMR90-HindIII-40K-tworeps.txt&#8221;
and &#8220;hitad-40K.log&#8221; will be generated under current working directory. We use a rotating file
for logging. According to our settings, when the size of &#8220;hitad-40K.log&#8221; gets about 200K, it&#8217;s
closed and renamed to &#8220;hitad-40K.log.1&#8221;. At the same time, a new file &#8220;hitad-40K.log&#8221; is silently
opened for output. In a word, the system saves old log files by appending the extensions
&#8221;.1&#8221;, &#8221;.2&#8221; etc., and the current log is always written to &#8220;hitad-40K.log&#8221;:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ head -6 IMR90-HindIII-40K-tworeps.txt

22      18240000        18560000        0
22      19000000        19640000        0
22      19920000        20120000        0
22      20800000        21480000        0
22      22040000        22320000        0
22      22320000        22720000        1
</pre></div>
</div>
<p>The output file contains 4 columns indicating chromosome label, domain start (bp), domain end (bp),
and hierarchical label, respectively. In our notation, TAD is denoted as level 0, sub-TAD is denoted
as level 1, and subsequent domain level is denoted as level 2, etc.</p>
</div>
<div class="section" id="transform-txt-into-npz">
<h4>Transform TXT into NPZ<a class="headerlink" href="#transform-txt-into-npz" title="Permalink to this headline">¶</a></h4>
<p>First, let&#8217;s take a look at a NPZ file within Python environment:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd ../lib/hitad/data/NPZ
</pre></div>
</div>
<p>Open a Python interperter and type:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lib</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;GM12878-HindIII.40K.rep1.npz&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lib</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[&#39;22&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">lib</span><span class="p">[</span><span class="s1">&#39;22&#39;</span><span class="p">])</span>
<span class="go">&lt;type &#39;numpy.ndarray&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lib</span><span class="p">[</span><span class="s1">&#39;22&#39;</span><span class="p">][:</span><span class="mi">5</span><span class="p">]</span>
<span class="go">array([(401, 686, 0.149032), (401, 693, 0.175786), (401, 695, 0.122059),</span>
<span class="go">       (401, 868, 0.156891), (401, 1218, 0.211272)],</span>
<span class="go">      dtype=[(&#39;bin1&#39;, &#39;&lt;i8&#39;), (&#39;bin2&#39;, &#39;&lt;i8&#39;), (&#39;IF&#39;, &#39;&lt;f8&#39;)])</span>
</pre></div>
</div>
<p>A NPZ file is a pool of arrays using chromosome labels as the keys. Each array is structured and has
3 fields: &#8220;bin1&#8221;, &#8220;bin2&#8221; and &#8220;IF&#8221;, just as the information contained in a TXT file mentioned
above.</p>
<p>NPZ files can be generated in two ways:</p>
<ol class="arabic simple">
<li>Run hitad with <code class="docutils literal"><span class="pre">--npzpre</span></code> specified.</li>
<li>Use <a class="reference external" href="https://github.com/XiaoTaoWang/HiC_pipeline">runHiC</a> pipeline. runHiC is a user-friendly
command-line software developed by our lab for Hi-C data processing.</li>
</ol>
<p>Let&#8217;s rerun hitad with <code class="docutils literal"><span class="pre">--npzpre</span></code> and see what will happen:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd ../../../../working/
$ hitad -O IMR90-HindIII-40K-tworeps.txt -d ../lib/hitad/data/TXT/datasets-tworeps --npzpre IMR90-HindIII --logFile hitad-40K.log
</pre></div>
</div>
<ul>
<li><p class="first"><code class="docutils literal"><span class="pre">--npzpre</span></code> NPZPRE</p>
<p>Prefix of output .npz file name. If no path is contained, NPZ files will be placed under
current working directory.</p>
</li>
</ul>
<p>Two additional files &#8220;IMR90-HindIII.40K.rep1.npz&#8221; and &#8220;IMR90-HindIII.40K.rep2.npz&#8221; will
be generated under current working directory for two replicates respectively.</p>
</div>
<div class="section" id="run-hitad-with-npz-data">
<h4>Run HiTAD with NPZ Data<a class="headerlink" href="#run-hitad-with-npz-data" title="Permalink to this headline">¶</a></h4>
<p>A metadata file is also required in the case of NPZ data:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd ../lib/hitad/data/NPZ
$ cat datasets-tworeps

res:40000
  rep1:../lib/hitad/data/NPZ/GM12878-HindIII.40K.rep1.npz
  rep2:../lib/hitad/data/NPZ/GM12878-HindIII.40K.rep2.npz
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Data path should point to the npz file.</p>
</div>
<p>Execute following commands and try to understand the result by yourself:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd ../../../../working/
$ hitad -O GM12878-HindIII-40K-tworeps.txt -d ../lib/hitad/data/NPZ/datasets-tworeps --logFile hitad-40K.log
</pre></div>
</div>
</div>
<div class="section" id="other-options">
<h4>Other Options<a class="headerlink" href="#other-options" title="Permalink to this headline">¶</a></h4>
<ul>
<li><p class="first"><code class="docutils literal"><span class="pre">-C/--chroms</span></code> [CHROMS [CHROMS ...]]</p>
<p>List of chromosome labels. Only Hi-C data within the specified chromosomes will be included.
Specially, &#8220;#&#8221; stands for chromosomes with numerical labels. The default setting is <code class="docutils literal"><span class="pre">--chrom</span> <span class="pre">&quot;#&quot;</span> <span class="pre">X</span></code>,
which will include 1, 2, ..., 22 and X chromosome for human genome. (Default: [&#8216;#&#8217;, &#8216;X&#8217;])</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--maxsize</span></code> MAXSIZE</p>
<p>Maximum domain size in base-pair unit. (Default: 4000000)</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">--removeCache</span></code></p>
<p>Remove cached data before existing. (Default: False)</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">-p/--cpu-core</span></code> CPU_CORE</p>
<p>Number of processes to launch. (default: 1)</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">-v/--version</span></code></p>
<p>Print version number and exit.</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">-h/--help</span></code></p>
<p>Show help message and exit.</p>
</li>
</ul>
</div>
</div>
<div class="section" id="domain-alignment">
<h3>Domain Alignment<a class="headerlink" href="#domain-alignment" title="Permalink to this headline">¶</a></h3>
<p>Traditionally, domains and boundaries are aligned by matching boundaries with nearest ones
between two data sets. This strategy generally assign a threshold in advance to determine
whether a boundary of one data set exists in the other data set. However, the selection
of the threshold is quite empirical and artificial. To deal with this problem, we propose
a parameter-free alignment strategy by maximizing the overlap ratio between matched domains
and considering all domains at the same time. We also generalize the strategy on hierarchical
TADs and further define change types at both boundary and domain levels systematically.</p>
<p>All related classes and functions are defined in <code class="xref py py-mod docutils literal"><span class="pre">tadlib.hitad.aligner</span></code>.</p>
<p>In this section, I will show you how to perform alignment/comparison between two
domain sets.</p>
<p>During the previous sections, we have generated 2 domain sets, &#8220;GM12878-HindIII-40K-tworeps.txt&#8221;
and &#8220;IMR90-HindIII-40K-tworeps.txt&#8221;, under the &#8220;working&#8221; directory.</p>
<p>Open a Python interperter and load these domain lists by
<a class="reference internal" href="hitad_api.html#tadlib.hitad.aligner.readHierDomain" title="tadlib.hitad.aligner.readHierDomain"><code class="xref py py-func docutils literal"><span class="pre">tadlib.hitad.aligner.readHierDomain()</span></code></a>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tadlib.hitad.aligner</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ilist</span> <span class="o">=</span> <span class="n">readHierDomain</span><span class="p">(</span><span class="s1">&#39;IMR90-HindIII-40K-tworeps.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">glist</span> <span class="o">=</span> <span class="n">readHierDomain</span><span class="p">(</span><span class="s1">&#39;GM12878-HindIII-40K-tworeps.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">ilist</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">glist</span><span class="p">)</span>
<span class="go">81 75</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">ilist</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">[[&#39;22&#39;, 18240000, 18560000, 0], [&#39;22&#39;, 19000000, 19640000, 0], [&#39;22&#39;, 19920000, 20120000, 0]]</span>
</pre></div>
</div>
<p>We also provide a function <a class="reference internal" href="hitad_api.html#tadlib.hitad.aligner.hierFormat" title="tadlib.hitad.aligner.hierFormat"><code class="xref py py-func docutils literal"><span class="pre">tadlib.hitad.aligner.hierFormat()</span></code></a> to transform
[chrom,start,end] domains to [chrom,start,end,level] domains. That will be helpful if
you want to align domain sets generated by other domain callers (such as <a class="reference external" href="https://github.com/theaidenlab/juicer/wiki/Arrowhead">Arrowhead</a>
and <a class="reference external" href="http://compbio.cs.brown.edu/projects/tadtree/">TADtree</a>).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pseudo</span> <span class="o">=</span> <span class="p">[</span><span class="n">d</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">ilist</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">pseudo</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">[[&#39;22&#39;, 18240000, 18560000], [&#39;22&#39;, 19000000, 19640000], [&#39;22&#39;, 19920000, 20120000]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">transformed</span> <span class="o">=</span> <span class="n">hierFormat</span><span class="p">(</span><span class="n">pseudo</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">transformed</span> <span class="o">==</span> <span class="n">ilist</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Hold domain lists with <a class="reference internal" href="hitad_api.html#tadlib.hitad.aligner.DomainSet" title="tadlib.hitad.aligner.DomainSet"><code class="xref py py-class docutils literal"><span class="pre">tadlib.hitad.aligner.DomainSet</span></code></a>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">iset</span> <span class="o">=</span> <span class="n">DomainSet</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span> <span class="n">ilist</span><span class="p">,</span> <span class="mi">40000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gset</span> <span class="o">=</span> <span class="n">DomainSet</span><span class="p">(</span><span class="s1">&#39;GM12878&#39;</span><span class="p">,</span> <span class="n">glist</span><span class="p">,</span> <span class="mi">40000</span><span class="p">)</span>
</pre></div>
</div>
<p>To construct hierarchical domain mapping between two domain sets, we need
to initialize a <a class="reference internal" href="hitad_api.html#tadlib.hitad.aligner.DomainAligner" title="tadlib.hitad.aligner.DomainAligner"><code class="xref py py-class docutils literal"><span class="pre">tadlib.hitad.aligner.DomainAligner</span></code></a> and invoke the
<a class="reference internal" href="hitad_api.html#tadlib.hitad.aligner.DomainAligner.align" title="tadlib.hitad.aligner.DomainAligner.align"><code class="xref py py-meth docutils literal"><span class="pre">tadlib.hitad.aligner.DomainAligner.align()</span></code></a> method:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">work</span> <span class="o">=</span> <span class="n">DomainAligner</span><span class="p">(</span><span class="n">iset</span><span class="p">,</span> <span class="n">gset</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">work</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p><a class="reference internal" href="hitad_api.html#tadlib.hitad.aligner.DomainAligner" title="tadlib.hitad.aligner.DomainAligner"><code class="xref py py-class docutils literal"><span class="pre">tadlib.hitad.aligner.DomainAligner</span></code></a> also defines domain-level change types, including
conserved TADs, semi-conserved TADs, merged TADs and split TADs:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">conserved</span> <span class="o">=</span> <span class="n">work</span><span class="o">.</span><span class="n">conserved</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span> <span class="c1"># Conserved TADs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">conserved</span><span class="p">)</span>
<span class="go">7</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">conserved</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># The first conserved TAD pair</span>
<span class="go">((&#39;22&#39;, 19000000, 19640000), (&#39;22&#39;, 19160000, 19680000))</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">semi</span> <span class="o">=</span> <span class="n">work</span><span class="o">.</span><span class="n">inner_changed</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span> <span class="c1"># Semi-Conserved TADs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">semi</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">semi</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># The first semi-conserved TAD pair</span>
<span class="go">((&#39;22&#39;, 27080000, 28240000), (&#39;22&#39;, 27080000, 28200000))</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">merged</span> <span class="o">=</span> <span class="n">work</span><span class="o">.</span><span class="n">merged</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span> <span class="c1"># Merged TADs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">merged</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># The first merged region</span>
<span class="go">((&#39;22&#39;, 29680000, 32360000), (&#39;22&#39;, 29680000, 32360000))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">merged</span><span class="p">[((</span><span class="s1">&#39;22&#39;</span><span class="p">,</span> <span class="mi">29680000</span><span class="p">,</span> <span class="mi">32360000</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;22&#39;</span><span class="p">,</span> <span class="mi">29680000</span><span class="p">,</span> <span class="mi">32360000</span><span class="p">))]</span> <span class="c1"># Merged details</span>
<span class="go">[[[&#39;22&#39;, 29680000, 31720000, 0], [&#39;22&#39;, 31720000, 32040000, 0], [&#39;22&#39;, 32040000, 32360000, 0]], [[&#39;22&#39;, 29680000, 32360000, 0]]]</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">split</span> <span class="o">=</span> <span class="n">work</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span> <span class="c1"># Split TADs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">split</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># The first split region</span>
<span class="go">((&#39;22&#39;, 20800000, 21480000), (&#39;22&#39;, 20840000, 21480000))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">split</span><span class="p">[((</span><span class="s1">&#39;22&#39;</span><span class="p">,</span> <span class="mi">20800000</span><span class="p">,</span> <span class="mi">21480000</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;22&#39;</span><span class="p">,</span> <span class="mi">20840000</span><span class="p">,</span> <span class="mi">21480000</span><span class="p">))]</span> <span class="c1"># Split details</span>
<span class="go">[[[&#39;22&#39;, 20800000, 21480000, 0]], [[&#39;22&#39;, 20840000, 21200000, 0], [&#39;22&#39;, 21200000, 21480000, 0]]]</span>
</pre></div>
</div>
<p>Boundary-level change types are defined in <a class="reference internal" href="hitad_api.html#tadlib.hitad.aligner.BoundAligner" title="tadlib.hitad.aligner.BoundAligner"><code class="xref py py-class docutils literal"><span class="pre">tadlib.hitad.aligner.BoundAligner</span></code></a> which
is also based on our domain-based alignment algorithm:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">boundview</span> <span class="o">=</span> <span class="n">BoundAligner</span><span class="p">(</span><span class="n">iset</span><span class="p">,</span> <span class="n">gset</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Pairs of conserved TAD boundaries</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">conserved_bounds</span> <span class="o">=</span> <span class="n">boundview</span><span class="o">.</span><span class="n">conserved_tad_bounds</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">conserved_bounds</span><span class="p">)[:</span><span class="mi">3</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span> <span class="n">i</span><span class="p">,</span> <span class="s1">&#39;&lt;--&gt;&#39;</span><span class="p">,</span> <span class="n">conserved_bounds</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<span class="gp">...</span>
<span class="go">(&#39;22&#39;, 19000000) &lt;--&gt; (&#39;22&#39;, 19160000)</span>
<span class="go">(&#39;22&#39;, 19640000) &lt;--&gt; (&#39;22&#39;, 19680000)</span>
<span class="go">(&#39;22&#39;, 19920000) &lt;--&gt; (&#39;22&#39;, 19920000)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Pairs of conserved sub-TAD boundaries.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">conserved_subs</span> <span class="o">=</span> <span class="n">boundview</span><span class="o">.</span><span class="n">conserved_sub_bounds</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">conserved_subs</span><span class="p">)[:</span><span class="mi">3</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span> <span class="n">i</span><span class="p">,</span> <span class="s1">&#39;&lt;--&gt;&#39;</span><span class="p">,</span> <span class="n">conserved_subs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<span class="gp">...</span>
<span class="go">(&#39;22&#39;, 26200000) &lt;--&gt; (&#39;22&#39;, 26200000)</span>
<span class="go">(&#39;22&#39;, 29200000) &lt;--&gt; (&#39;22&#39;, 29240000)</span>
<span class="go">(&#39;22&#39;, 29960000) &lt;--&gt; (&#39;22&#39;, 29960000)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># TAD to sub-TAD switch cases</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tad_sub</span> <span class="o">=</span> <span class="n">boundview</span><span class="o">.</span><span class="n">tad2sub</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">tad_sub</span><span class="p">)</span>
<span class="go">7</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># sub-TAD to TAD switch cases.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sub_tad</span> <span class="o">=</span> <span class="n">boundview</span><span class="o">.</span><span class="n">sub2tad</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">sub_tad</span><span class="p">)</span>
<span class="go">8</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># TAD boundaries that exist in IMR90, but disappear in GM12878.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">disappear_TAD</span> <span class="o">=</span> <span class="n">boundview</span><span class="o">.</span><span class="n">disappeared_tad</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">disappear_TAD</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># The first disappeared TAD boundary</span>
<span class="go">(&#39;22&#39;, 18240000)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Sub-TAD boundaries that exist in IMR90, but disappear in GM12878.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">disappear_sub</span> <span class="o">=</span> <span class="n">boundview</span><span class="o">.</span><span class="n">disappeared_sub</span><span class="p">(</span><span class="s1">&#39;IMR90&#39;</span><span class="p">,</span><span class="s1">&#39;GM12878&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">disappear_sub</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># The first disappeared sub-TAD boundary</span>
<span class="go">(&#39;22&#39;, 23480000)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="api-documentation">
<h2>API Documentation<a class="headerlink" href="#api-documentation" title="Permalink to this headline">¶</a></h2>
<p>API reference of our defined classes and functions for hierarchical TAD
identification and domain alignment.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="hitad_api.html">Hierarchical TAD Identification</a></li>
<li class="toctree-l1"><a class="reference internal" href="hitad_api.html#domain-loading-and-aligning">Domain Loading and Aligning</a></li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, XiaoTao Wang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/hitad.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/XiaoTaoWang/TADLib" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>